### PR TITLE
feat(ci): merge gate enforces fix-bot pinning-test discipline

### PR DESCRIPTION
**Pull request by kilbot**
Mechanical backstop for the fleet's Pinning-Test Discipline (companion: wcpos/wcpos-openclaw#1534 — contract + eval pins; context in wcpos/woocommerce-pos#1310 and wcpos/monorepo#790, where three fixer commits in one day shipped behavior changes without their proof).

## What

New `enforce_bot_fix_discipline` check in `merge-gate.sh`: any PR commit authored by a fix bot (`MERGE_GATE_FIX_BOT_AUTHORS`, default `wcpos-agents[bot]`) that changes source files must:

1. touch a **test file in the same commit** (a fix is not a fix until a test pins it), and
2. carry a **`Tested:` trailer** recording the local suite run and environment.

Fails closed with a message naming the offending commit and the missing half. Human commits and docs-only bot commits are exempt. Runs before the translation/POT allowlist bypasses, so no PR shape skips it. API lookup failures fail closed.

## Targeting main

The gate workflow checks out the **default branch's** scripts (trusted-base pattern), so landing this on `main` makes it govern PRs on every lane immediately — no next-lane port needed.

## Tests

4 new harness cases in `test-merge-gate.sh` (mock `gh api` extended for commit listing/inspection): source-without-test fails, test-without-trailer fails, test+trailer passes (with a human commit correctly exempt in the same PR), docs-only bot commit exempt. Full harness green.

Tested: `bash .github/scripts/test-merge-gate.sh` — all cases pass.

## Summary by CodeRabbit

* **New Features**
  * Added pre-merge checks for automated fix commits to ensure source changes include a corresponding test update.
  * Added validation that fix commits include a meaningful `Tested:` result in the commit message.
  * Checks now fail safely when required commit or file information cannot be retrieved.

* **Tests**
  * Expanded merge-gate coverage for missing tests, invalid testing details, test deletions, and related edge cases.

### CODE REVIEW

File: .github/scripts/merge-gate.sh [hunk 1]
```diff
@@ -182,6 +182,81 @@ wait_for_checks() {
...
+  case "$1" in
+    tests/*|*/tests/*|*.test.*|*.spec.*) return 0 ;;
+    *) return 1 ;;
+  esac
+}
+
+is_source_path() {
+  is_test_path "$1" && return 1
+  case "$1" in
+    *.php|*.ts|*.tsx|*.js|*.jsx) return 0 ;;
```
[the remark this answers is not in the document]
Comment (kilbot) [reply .github/scripts/merge-gate.sh]: Fixed in d712c68a — .mjs/.cjs/.mts/.cts now count as source (consistency port from the sibling review on wcpos/monorepo#792); harness case included.

File: .github/scripts/merge-gate.sh [hunk 2]
```diff
@@ -182,6 +182,81 @@ wait_for_checks() {
...
 
+FIX_BOT_AUTHORS="|${MERGE_GATE_FIX_BOT_AUTHORS:-wcpos-agents[bot]}|"
+
+pr_commits() {
+  gh api "repos/${GITHUB_REPOSITORY}/pulls/${PR_NUMBER}/commits" --paginate \
+    --jq '.[] | [.sha, (.author.login // .commit.author.name // "unknown")] | @tsv'
+}
+
+commit_files() {
+  gh api "repos/${GITHUB_REPOSITORY}/commits/$1" --jq '.files[].filename'
```
[the remark this answers is not in the document]
Comment (kilbot) [reply .github/scripts/merge-gate.sh]: Fixed in d712c68a — commit files now carry their status ('.files[] | [.status, .filename] | @tsv') and 'removed' test paths are ignored, so deleting a test no longer satisfies the pin. Harness case included.

File: .github/scripts/merge-gate.sh [hunk 3]
```diff
@@ -182,6 +182,95 @@ wait_for_checks() {
...
+}
+
+# The Tested: line must sit in the message's FINAL paragraph (the git trailer
+# block) — prose that merely mentions "Tested:" mid-body does not count.
+trailer_block_has_tested() {
+  printf '%s\n' "$1" | awk '
+    BEGIN { block = "" }
+    /^[[:space:]]*$/ { block = ""; next }
+    { block = block $0 "\n" }
+    END { exit (block ~ /(^|\n)Tested:/) ? 0 : 1 }
```
[the remark this answers is not in the document]
Comment (kilbot) [reply .github/scripts/merge-gate.sh]: Fixed in eb8af6e9 — the trailer value must now be result-shaped: at least 8 chars containing a digit (real suite results quote counts), so 'Tested:', 'Tested: N/A', or a bare command no longer pass. Same tightening applied to the monorepo and pro gates; harness case added in each.

File: .github/scripts/merge-gate.sh [hunk 4]
```diff
@@ -182,6 +182,104 @@ wait_for_checks() {
...
+  # "Tested:", "Tested: N/A", or a command with no result do not count.
+  printf '%s\n' "$1" | awk '
+    BEGIN { block = "" }
+    /^[[:space:]]*$/ { block = ""; next }
+    { block = block $0 "\n" }
+    END {
+      if (match(block, /(^|\n)Tested:[^\n]*/) == 0) exit 1
+      value = substr(block, RSTART, RLENGTH)
+      sub(/(^|\n)Tested:[[:space:]]*/, "", value)
+      exit (length(value) >= 8 && value ~ /[0-9]/) ? 0 : 1
```
[the remark this answers is not in the document]
Comment (kilbot) [reply .github/scripts/merge-gate.sh:232]: Declined — this is where textual validation hits its ceiling, and tightening further trades real false positives for no real protection.

The gate is a tripwire for OMISSION, not a lie detector. All three motivating incidents (wcpos/woocommerce-pos#1310, wcpos/monorepo#790) involved agents that never ran the suite and therefore had no result line to paste — none involved a fabricated trailer. An agent willing to write 'Tested: 12345678' is lying, and no bash regex can distinguish a well-formed lie from a real result; that class is handled by the layers that own intent: the sidecar contract + evals (wcpos/wcpos-openclaw#1534) governing what fix runs write, and CI itself — the suite runs on the PR regardless, so a forged trailer over broken code still fails the checks. The trailer's job is to prove the agent looked before pushing, which the shape check (digit + substance, in the trailer block) already forces.

Meanwhile, legitimate result formats in this week's real commits alone vary widely: 'Tested: OK (79 tests, 334 assertions) — wp-env WC 10.4.3' (phpunit), 'Tested: 8/8 green — jest use-refund-mutation', 'Tested: OK (269 tests, 0 fail) — Node.js v22.23.1 agent-evals' (node --test), 'Tested: bash .github/scripts/test-merge-gate.sh — all cases pass incl. 4 new'. Any pattern strict enough to reject 'abc12345' starts rejecting some of these — a false positive here blocks a compliant fix and teaches agents the gate is noise.

File: .github/scripts/merge-gate.sh [hunk 5]
```diff
@@ -182,6 +182,104 @@ wait_for_checks() {
...
+      log "Could not read the message for fix-bot commit ${sha:0:8}; failing closed."
+      return 1
+    fi
+    if ! trailer_block_has_tested "$msg"; then
+      log "✗ Fix-bot commit ${sha:0:8} ($author) has no 'Tested:' trailer. Run the touched suite locally and record the literal result line (e.g. 'Tested: OK (79 tests) — wp-env WC 10.4.3')."
+      failed=1
+    fi
+  done <<< "$commits"
+  return "$failed"
+}
```
[the remark this answers is not in the document]
Comment (kilbot) [reply .github/scripts/merge-gate.sh:304]: Fixed in 27964de9 — at 300+ files the gate now fails closed with a split-it message rather than judging a truncated list: a commit that size violates the fleet's small-directed-fix contract regardless of what the hidden entries contain, so refusing outright is both simpler and stricter than paginating. Same treatment in the monorepo and pro gates; harness case added in each.

File: .github/scripts/merge-gate.sh [hunk 6]
```diff
@@ -182,6 +182,113 @@ wait_for_checks() {
...
+    tests/*|*/tests/*|*.test.*|*.spec.*|*/test-*.sh|test-*.sh) return 0 ;;
+    *) return 1 ;;
+  esac
+}
+
+is_source_path() {
+  is_test_path "$1" && return 1
+  case "$1" in
+    *.php|*.ts|*.tsx|*.js|*.jsx|*.mjs|*.cjs|*.mts|*.cts) return 0 ;;
+    .github/scripts/*.sh|scripts/*.sh) return 0 ;;
```
[the remark this answers is not in the document]
Comment (kilbot) [reply .github/scripts/merge-gate.sh:213]: Fixed in 9e59449f (mirrored to monorepo#793 and pro#381) — a new config class (.github/workflows/*, .github/*.json, composer/package manifests + lockfiles) closes the skip: config-only bot commits now require the Tested: trailer. Deliberately NOT the same-commit test: a failing-first test has no meaningful form for a version bump or workflow tweak, and this repo's own requires_php_tests encodes exactly that philosophy — config changes are proven by running the suite. Source-class requirements unchanged; two harness cases per repo.

One process note: this is review round 4 on this PR, each round converging on classifier completeness (source → tests → gate scripts → config). The classifier now partitions every path: test / source / config / exempt-docs. If a round 5 finds another class, I'll escalate to the maintainer for a scope decision rather than continue iterating.
